// @donkeyclip/motorcortex/src/Clip/Browser/Descriptive/helpers/NullClip.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { successWithEmptyExecution } from '../../../../Channels/Channel';

class NullClip {
  constructor() {
    this.isNullClip = true;
  }

  get duration() {
    return 0;
  }

  addIncident() {
    return successWithEmptyExecution;
  }

  // eslint-disable-next-line sonarjs/no-identical-functions
  moveIncident() {
    return successWithEmptyExecution;
  }

  // eslint-disable-next-line sonarjs/no-identical-functions
  removeIncident() {
    return successWithEmptyExecution;
  }

  // eslint-disable-next-line sonarjs/no-identical-functions
  resizeIncident() {
    return successWithEmptyExecution;
  }

  getIncidentsByChannel() {
    return {};
  }

  flash() {
    // this is left intentionally empty
  }

  _resize() {
    // this is left intentionally empty
  }

  onProgress() {
    // this is left intentionally empty
  }

  renderOnDOM() {
    // this is left intentionally empty
  }

  removeFromDOM() {
    // this is left intentionally empty
  }
}

export default NullClip;
