// @donkeyclip/motorcortex/src/DescriptiveIncidents/__tests__/states/combo_state_1.js
export const descriptiveTree = {
  id: 'my-clip',
  duration: 6700,
  children: {
    bombo: {
      id: 'bombo',
      position: 1000,
      children: {},
      duration: 5700,
    },
  },
};
export const lanes = {};
export const realTree = {
  id: 'my-clip',
  duration: 6900,
  children: {
    'desc-combo': {
      id: 'desc-combo',
      position: 1000, // ok
      duration: 5900, // ok
      children: {
        'desc-combo_element-0': {
          id: 'desc-combo_element-0',
          duration: 4500, // ok
          position: 0, // ok
          children: {
            'desc-combo_element-0_repeat-0': {
              id: 'desc-combo_element-0_repeat-0',
              duration: 2100, // ok
              position: 100, // ok
              children: {
                'desc-combo_element-0_repeat-0_incident-0': {
                  id: 'desc-combo_element-0_repeat-0_incident-0',
                  position: 0, // ok
                  duration: 1000, // ok
                  children: {
                    'desc-combo_element-0_repeat-0_incident-0_combo-target-1': {
                      id: 'desc-combo_element-0_repeat-0_incident-0_combo-target-1',
                      position: 0, // ok
                      duration: 1000, // ok
                      children: {
                        'desc-combo_element-0_repeat-0_incident-0_combo-target-1_alpha':
                          {
                            id: 'desc-combo_element-0_repeat-0_incident-0_combo-target-1_alpha',
                            position: 0, // ok
                            duration: 1000, // ok
                          },
                      },
                    },
                  },
                },
                'desc-combo_element-0_repeat-0_incident-1': {
                  id: 'desc-combo_element-0_repeat-0_incident-1',
                  position: 1000, // ok
                  duration: 1100, // ok
                  children: {
                    'desc-combo_element-0_repeat-0_incident-1_stagger-1-1': {
                      id: 'desc-combo_element-0_repeat-0_incident-1_stagger-1-1',
                      position: 0, // ok -- check
                      duration: 1000, // ok
                      children: {
                        'desc-combo_element-0_repeat-0_incident-1_stagger-1-1_alpha':
                          {
                            id: 'desc-combo_element-0_repeat-0_incident-1_stagger-1-1_alpha',
                            position: 0, // ok -- check
                            duration: 1000, // ok
                          },
                      },
                    },
                    'desc-combo_element-0_repeat-0_incident-1_stagger-1-2': {
                      id: 'desc-combo_element-0_repeat-0_incident-1_stagger-1-2',
                      position: 0, // ok
                      duration: 1100, // ok
                      children: {
                        'desc-combo_element-0_repeat-0_incident-1_stagger-1-2_alpha':
                          {
                            id: 'desc-combo_element-0_repeat-0_incident-1_stagger-1-2_alpha',
                            position: 0, // ok
                            duration: 1100, // ok
                          },
                      },
                    },
                  },
                },
              },
            },
            'desc-combo_element-0_repeat-1': {
              id: 'desc-combo_element-0_repeat-1',
              duration: 2100, // ok
              position: 2400, // ok
              children: {
                'desc-combo_element-0_repeat-1_incident-0': {
                  id: 'desc-combo_element-0_repeat-1_incident-0',
                  position: 0, // ok
                  duration: 1000, // ok
                  children: {
                    'desc-combo_element-0_repeat-1_incident-0_combo-target-1': {
                      id: 'desc-combo_element-0_repeat-1_incident-0_combo-target-1',
                      position: 0, // ok
                      duration: 1000, // ok
                      children: {
                        'desc-combo_element-0_repeat-1_incident-0_combo-target-1_alpha':
                          {
                            id: 'desc-combo_element-0_repeat-1_incident-0_combo-target-1_alpha',
                            position: 0, // ok
                            duration: 1000, // ok
                          },
                      },
                    },
                  },
                },
                'desc-combo_element-0_repeat-1_incident-1': {
                  id: 'desc-combo_element-0_repeat-1_incident-1',
                  position: 1000, // ok
                  duration: 1100, // ok
                  children: {
                    'desc-combo_element-0_repeat-1_incident-1_stagger-1-1': {
                      id: 'desc-combo_element-0_repeat-1_incident-1_stagger-1-1',
                      position: 0, // ok
                      duration: 1000, // ok
                      children: {
                        'desc-combo_element-0_repeat-1_incident-1_stagger-1-1_alpha':
                          {
                            id: 'desc-combo_element-0_repeat-1_incident-1_stagger-1-1_alpha',
                            position: 0, // ok
                            duration: 1000, // ok
                          },
                      },
                    },
                    'desc-combo_element-0_repeat-1_incident-1_stagger-1-2': {
                      id: 'desc-combo_element-0_repeat-1_incident-1_stagger-1-2',
                      position: 0, // ok
                      duration: 1100, // ok
                      children: {
                        'desc-combo_element-0_repeat-1_incident-1_stagger-1-2_alpha':
                          {
                            id: 'desc-combo_element-0_repeat-1_incident-1_stagger-1-2_alpha',
                            position: 0, // ok
                            duration: 1100, // ok
                          },
                      },
                    },
                  },
                },
              },
            },
          },
        },
        'desc-combo_element-1': {
          id: 'desc-combo_element-1',
          duration: 5900, // ok
          position: 0, // ok
          children: {
            'desc-combo_element-1_repeat-0': {
              id: 'desc-combo_element-1_repeat-0',
              duration: 2600, // ok
              position: 300, // ok
              children: {
                'desc-combo_element-1_repeat-0_incident-0': {
                  id: 'desc-combo_element-1_repeat-0_incident-0',
                  position: 100, // ok
                  duration: 1030, // ok -- check
                  children: {
                    'desc-combo_element-1_repeat-0_incident-0_combo-target-2': {
                      id: 'desc-combo_element-1_repeat-0_incident-0_combo-target-2',
                      position: 0, // ok
                      duration: 1030, // ok
                      children: {
                        'desc-combo_element-1_repeat-0_incident-0_combo-target-2_alpha':
                          {
                            id: 'desc-combo_element-1_repeat-0_incident-0_combo-target-2_alpha',
                            position: 0, // ok
                            duration: 1030, // ok
                          },
                      },
                    },
                  },
                },
                'desc-combo_element-1_repeat-0_incident-1': {
                  id: 'desc-combo_element-1_repeat-0_incident-1',
                  position: 1500, // ok
                  duration: 1100, // ok
                  children: {
                    'desc-combo_element-1_repeat-0_incident-1_stagger-2-1': {
                      id: 'desc-combo_element-1_repeat-0_incident-1_stagger-2-1',
                      position: 0, // ok
                      duration: 1000, // ok
                      children: {
                        'desc-combo_element-1_repeat-0_incident-1_stagger-2-1_alpha':
                          {
                            id: 'desc-combo_element-1_repeat-0_incident-1_stagger-2-1_alpha',
                            position: 0, // ok
                            duration: 1000, // ok
                          },
                      },
                    },
                    'desc-combo_element-1_repeat-0_incident-1_stagger-2-2': {
                      id: 'desc-combo_element-1_repeat-0_incident-1_stagger-2-2',
                      position: 0, // ok
                      duration: 1100, // ok
                      children: {
                        'desc-combo_element-1_repeat-0_incident-1_stagger-2-2_alpha':
                          {
                            id: 'desc-combo_element-1_repeat-0_incident-1_stagger-2-2_alpha',
                            position: 0, // ok
                            duration: 1100, // ok
                          },
                      },
                    },
                  },
                },
              },
            },
            'desc-combo_element-1_repeat-1': {
              id: 'desc-combo_element-1_repeat-1',
              duration: 2600, // ok
              position: 3300, // ok
              children: {
                'desc-combo_element-1_repeat-1_incident-0': {
                  id: 'desc-combo_element-1_repeat-1_incident-0',
                  position: 100, // ok
                  duration: 1030, // ok
                  children: {
                    'desc-combo_element-1_repeat-1_incident-0_combo-target-2': {
                      id: 'desc-combo_element-1_repeat-1_incident-0_combo-target-2',
                      position: 0, // ok
                      duration: 1030, // ok
                      children: {
                        'desc-combo_element-1_repeat-1_incident-0_combo-target-2_alpha':
                          {
                            id: 'desc-combo_element-1_repeat-1_incident-0_combo-target-2_alpha',
                            position: 0, // ok
                            duration: 1030, // ok
                          },
                      },
                    },
                  },
                },
                'desc-combo_element-1_repeat-1_incident-1': {
                  id: 'desc-combo_element-1_repeat-1_incident-1',
                  position: 1500, // ok
                  duration: 1100, // ok
                  children: {
                    'desc-combo_element-1_repeat-1_incident-1_stagger-2-1': {
                      id: 'desc-combo_element-1_repeat-1_incident-1_stagger-2-1',
                      position: 0, // ok
                      duration: 1000, // ok
                      children: {
                        'desc-combo_element-1_repeat-1_incident-1_stagger-2-1_alpha':
                          {
                            id: 'desc-combo_element-1_repeat-1_incident-1_stagger-2-1_alpha',
                            position: 0, // ok
                            duration: 1000, // ok
                          },
                      },
                    },
                    'desc-combo_element-1_repeat-1_incident-1_stagger-2-2': {
                      id: 'desc-combo_element-1_repeat-1_incident-1_stagger-2-2',
                      position: 0, // ok
                      duration: 1100, // ok
                      children: {
                        'desc-combo_element-1_repeat-1_incident-1_stagger-2-2_alpha':
                          {
                            id: 'desc-combo_element-1_repeat-1_incident-1_stagger-2-2_alpha',
                            position: 0,
                            duration: 1100,
                          },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
};
