// @donkeyclip/motorcortex/src/DescriptiveIncidents/__tests__/states/dc25.js
export const descriptiveTree = {
  id: 'my-clip',
  duration: 8000 + 6 * (2000 + 1000 + 50),
  children: {
    'my-group': {
      id: 'my-group',
      position: 0,
      duration: 2400,
      children: {
        'test-incident-5': {
          id: 'test-incident-5',
          position: 0,
          duration: 2400,
        },
      },
    },
    'stagger-attributes-incident': {
      id: 'stagger-attributes-incident',
      position: 2000,
      duration: 800,
    },
    'test-incident-8': {
      id: 'test-incident-8',
      position: 6000,
      duration: 1000,
    },
    'test-incident-9': {
      id: 'test-incident-9',
      position: 8000,
      duration: 6 * (2000 + 1000 + 50),
    },
    'test-incident-2': {
      id: 'test-incident-2',
      position: 5000,
      duration: 6000,
    },
    'my-plugin-group': {
      id: 'my-plugin-group',
      position: 4000,
      duration: 300,
      children: {
        'my-plugin-group_0': {
          id: 'my-plugin-group_0',
          duration: 100,
          position: 0,
        },
        'my-plugin-group_1': {
          id: 'my-plugin-group_1',
          duration: 100,
          position: 100,
        },
        'my-plugin-group_2': {
          id: 'my-plugin-group_2',
          duration: 100,
          position: 200,
        },
      },
    },
    'test-group-1': {
      id: 'test-group-1',
      position: 1000,
      duration: 9000,
      children: {
        'test-incident-1': {
          id: 'test-incident-1',
          position: 7000,
          duration: 2000,
        },
      },
    },
  },
};
export const lanes = {
  'my-class-1___alpha': [
    {
      id: 'test-incident-5_my-class-1_alpha',
      millisecond: 0,
      initialValue: 1,
    },
    {
      id: 'my-plugin-group_0_my-class-1_alpha',
      millisecond: 4000,
      initialValue: 5,
    },
    {
      id: 'my-plugin-group_1_my-class-1_alpha',
      millisecond: 4100,
      initialValue: 2,
    },
    {
      id: 'my-plugin-group_2_my-class-1_alpha',
      millisecond: 4200,
      initialValue: 0,
    },
    {
      id: 'test-incident-1_my-class-1_alpha',
      millisecond: 8000,
      initialValue: 2,
    },
  ],
  'my-class-1___beta': [
    {
      id: 'test-incident-5_my-class-1_beta',
      millisecond: 0,
      initialValue: 0,
    },
    {
      id: 'test-incident-1_my-class-1_beta',
      millisecond: 8000,
      initialValue: 5,
    },
  ],
  'my-class-1___gamma': [
    {
      id: 'test-incident-2_my-class-1_gamma',
      millisecond: 5000,
      initialValue: 3,
    },
  ],
  'my-class-2___alpha': [
    {
      id: 'test-incident-5_my-class-2_alpha',
      millisecond: 0,
      initialValue: 1,
    },
    {
      id: 'test-incident-1_my-class-2_alpha',
      millisecond: 8000,
      initialValue: 5,
    },
  ],
  'my-class-2___beta': [
    {
      id: 'test-incident-5_my-class-2_beta',
      millisecond: 0,
      initialValue: 0,
    },
    {
      id: 'test-incident-1_my-class-2_beta',
      millisecond: 8000,
      initialValue: 5,
    },
  ],
  'stagger-1___alpha': [
    {
      id: 'stagger-attributes-incident_stagger-1_alpha',
      millisecond: 2000,
      initialValue: 1,
    },
    {
      id: 'test-incident-8_stagger-1_alpha',
      millisecond: 6000,
      initialValue: 0,
    },
    {
      id: 'test-incident-9_stagger-1_alpha',
      millisecond: 8000,
      initialValue: 6,
    },
  ],
  'stagger-2___alpha': [
    {
      id: 'stagger-attributes-incident_stagger-2_alpha',
      millisecond: 2000,
      initialValue: 1,
    },
    {
      id: 'test-incident-8_stagger-2_alpha',
      millisecond: 6000,
      initialValue: 200,
    },
    {
      id: 'test-incident-9_stagger-2_alpha',
      millisecond: 8000,
      initialValue: 6,
    },
  ],
  'stagger-3___alpha': [
    {
      id: 'stagger-attributes-incident_stagger-3_alpha',
      millisecond: 2000,
      initialValue: 1,
    },
    {
      id: 'test-incident-8_stagger-3_alpha',
      millisecond: 6000,
      initialValue: 400,
    },
    {
      id: 'test-incident-9_stagger-3_alpha',
      millisecond: 8000,
      initialValue: 6,
    },
  ],
  'stagger-4___alpha': [
    {
      id: 'stagger-attributes-incident_stagger-4_alpha',
      millisecond: 2000,
      initialValue: 1,
    },
    {
      id: 'test-incident-8_stagger-4_alpha',
      millisecond: 6000,
      initialValue: 600,
    },
    {
      id: 'test-incident-9_stagger-4_alpha',
      millisecond: 8000,
      initialValue: 6,
    },
  ],
  'stagger-5___alpha': [
    {
      id: 'stagger-attributes-incident_stagger-5_alpha',
      millisecond: 2000,
      initialValue: 1,
    },
    {
      id: 'test-incident-8_stagger-5_alpha',
      millisecond: 6000,
      initialValue: 800,
    },
    {
      id: 'test-incident-9_stagger-5_alpha',
      millisecond: 8000,
      initialValue: 6,
    },
  ],
  'stagger-6___alpha': [
    {
      id: 'stagger-attributes-incident_stagger-6_alpha',
      millisecond: 2000,
      initialValue: 1,
    },
    {
      id: 'test-incident-8_stagger-6_alpha',
      millisecond: 6000,
      initialValue: 1000,
    },
    {
      id: 'test-incident-9_stagger-6_alpha',
      millisecond: 8000,
      initialValue: 6,
    },
  ],
};
export const realTree = {
  id: 'my-clip',
  duration: 8000 + 6 * (2000 + 1000 + 50),
  children: {
    'my-group': {
      id: 'my-group',
      position: 0,
      duration: 2400,
      children: {
        'test-incident-5': {
          id: 'test-incident-5',
          position: 0,
          duration: 2400,
          children: {
            'test-incident-5_my-class-1': {
              id: 'test-incident-5_my-class-1',
              position: 0,
              duration: 2400,
              children: {
                'test-incident-5_my-class-1_alpha': {
                  id: 'test-incident-5_my-class-1_alpha',
                  position: 0,
                  duration: 2400,
                },
                'test-incident-5_my-class-1_beta': {
                  id: 'test-incident-5_my-class-1_beta',
                  position: 0,
                  duration: 2400,
                },
              },
            },
            'test-incident-5_my-class-2': {
              id: 'test-incident-5_my-class-2',
              position: 0,
              duration: 2400,
              children: {
                'test-incident-5_my-class-2_alpha': {
                  id: 'test-incident-5_my-class-2_alpha',
                  position: 0,
                  duration: 2400,
                },
                'test-incident-5_my-class-2_beta': {
                  id: 'test-incident-5_my-class-2_beta',
                  position: 0,
                  duration: 2400,
                },
              },
            },
          },
        },
      },
    },
    'my-plugin-group': {
      id: 'my-plugin-group',
      position: 4000,
      duration: 300,
      children: {
        'my-plugin-group_0': {
          id: 'my-plugin-group_0',
          position: 0,
          duration: 100,
          children: {
            'my-plugin-group_0_my-class-1': {
              id: 'my-plugin-group_0_my-class-1',
              position: 0,
              duration: 100,
              children: {
                'my-plugin-group_0_my-class-1_alpha': {
                  id: 'my-plugin-group_0_my-class-1_alpha',
                  position: 0,
                  duration: 100,
                },
              },
            },
          },
        },
        'my-plugin-group_1': {
          id: 'my-plugin-group_1',
          position: 100,
          duration: 100,
          children: {
            'my-plugin-group_1_my-class-1': {
              id: 'my-plugin-group_1_my-class-1',
              position: 0,
              duration: 100,
              children: {
                'my-plugin-group_1_my-class-1_alpha': {
                  id: 'my-plugin-group_1_my-class-1_alpha',
                  position: 0,
                  duration: 100,
                },
              },
            },
          },
        },
        'my-plugin-group_2': {
          id: 'my-plugin-group_2',
          position: 200,
          duration: 100,
          children: {
            'my-plugin-group_2_my-class-1': {
              id: 'my-plugin-group_2_my-class-1',
              position: 0,
              duration: 100,
              children: {
                'my-plugin-group_2_my-class-1_alpha': {
                  id: 'my-plugin-group_2_my-class-1_alpha',
                  position: 0,
                  duration: 100,
                },
              },
            },
          },
        },
      },
    },
    'test-incident-2': {
      id: 'test-incident-2',
      position: 5000,
      duration: 6000,
      children: {
        'test-incident-2_my-class-1': {
          id: 'test-incident-2_my-class-1',
          position: 0,
          duration: 6000,
          children: {
            'test-incident-2_my-class-1_gamma': {
              id: 'test-incident-2_my-class-1_gamma',
              position: 0,
              duration: 6000,
            },
          },
        },
      },
    },

    'test-incident-9': {
      id: 'test-incident-9',
      position: 8000,
      duration: 6 * (2000 + 1000 + 50),
      children: {
        'test-incident-9_stagger-1': {
          id: 'test-incident-9_stagger-1',
          position: 0,
          duration: 1 * (1000 + 0 + 0),
          children: {
            'test-incident-9_stagger-1_alpha': {
              id: 'test-incident-9_stagger-1_alpha',
              position: 0,
              duration: 1 * (1000 + 0 + 0),
            },
          },
        },
        'test-incident-9_stagger-2': {
          id: 'test-incident-9_stagger-2',
          position: 0,
          duration: 2 * (1200 + 200 + 10),
          children: {
            'test-incident-9_stagger-2_alpha': {
              id: 'test-incident-9_stagger-2_alpha',
              position: 0,
              duration: 2 * (1200 + 200 + 10),
            },
          },
        },
        'test-incident-9_stagger-3': {
          id: 'test-incident-9_stagger-3',
          position: 0,
          duration: 3 * (1400 + 400 + 20),
          children: {
            'test-incident-9_stagger-3_alpha': {
              id: 'test-incident-9_stagger-3_alpha',
              position: 0,
              duration: 3 * (1400 + 400 + 20),
            },
          },
        },
        'test-incident-9_stagger-4': {
          id: 'test-incident-9_stagger-4',
          position: 0,
          duration: 4 * (1600 + 600 + 30),
          children: {
            'test-incident-9_stagger-4_alpha': {
              id: 'test-incident-9_stagger-4_alpha',
              position: 0,
              duration: 4 * (1600 + 600 + 30),
            },
          },
        },
        'test-incident-9_stagger-5': {
          id: 'test-incident-9_stagger-5',
          position: 0,
          duration: 5 * (1800 + 800 + 40),
          children: {
            'test-incident-9_stagger-5_alpha': {
              id: 'test-incident-9_stagger-5_alpha',
              position: 0,
              duration: 5 * (1800 + 800 + 40),
            },
          },
        },
        'test-incident-9_stagger-6': {
          id: 'test-incident-9_stagger-6',
          position: 0,
          duration: 6 * (2000 + 1000 + 50),
          children: {
            'test-incident-9_stagger-6_alpha': {
              id: 'test-incident-9_stagger-6_alpha',
              position: 0,
              duration: 6 * (2000 + 1000 + 50),
            },
          },
        },
      },
    },

    'test-incident-8': {
      id: 'test-incident-8',
      position: 6000,
      duration: 1000,
      children: {
        'test-incident-8_stagger-1': {
          id: 'test-incident-8_stagger-1',
          position: 0,
          duration: 1000,
          children: {
            'test-incident-8_stagger-1_alpha': {
              id: 'test-incident-8_stagger-1_alpha',
              position: 0,
              duration: 1000,
            },
          },
        },
        'test-incident-8_stagger-2': {
          id: 'test-incident-8_stagger-2',
          position: 0,
          duration: 1000,
          children: {
            'test-incident-8_stagger-2_alpha': {
              id: 'test-incident-8_stagger-2_alpha',
              position: 0,
              duration: 1000,
            },
          },
        },
        'test-incident-8_stagger-3': {
          id: 'test-incident-8_stagger-3',
          position: 0,
          duration: 1000,
          children: {
            'test-incident-8_stagger-3_alpha': {
              id: 'test-incident-8_stagger-3_alpha',
              position: 0,
              duration: 1000,
            },
          },
        },
        'test-incident-8_stagger-4': {
          id: 'test-incident-8_stagger-4',
          position: 0,
          duration: 1000,
          children: {
            'test-incident-8_stagger-4_alpha': {
              id: 'test-incident-8_stagger-4_alpha',
              position: 0,
              duration: 1000,
            },
          },
        },
        'test-incident-8_stagger-5': {
          id: 'test-incident-8_stagger-5',
          position: 0,
          duration: 1000,
          children: {
            'test-incident-8_stagger-5_alpha': {
              id: 'test-incident-8_stagger-5_alpha',
              position: 0,
              duration: 1000,
            },
          },
        },
        'test-incident-8_stagger-6': {
          id: 'test-incident-8_stagger-6',
          position: 0,
          duration: 1000,
          children: {
            'test-incident-8_stagger-6_alpha': {
              id: 'test-incident-8_stagger-6_alpha',
              position: 0,
              duration: 1000,
            },
          },
        },
      },
    },
    'stagger-attributes-incident': {
      id: 'stagger-attributes-incident',
      position: 2000,
      duration: 800,
      children: {
        'stagger-attributes-incident_stagger-1': {
          id: 'stagger-attributes-incident_stagger-1',
          position: 0,
          duration: 800,
          children: {
            'stagger-attributes-incident_stagger-1_alpha': {
              id: 'stagger-attributes-incident_stagger-1_alpha',
              position: 0,
              duration: 800,
            },
          },
        },
        'stagger-attributes-incident_stagger-2': {
          id: 'stagger-attributes-incident_stagger-2',
          position: 0,
          duration: 800,
          children: {
            'stagger-attributes-incident_stagger-2_alpha': {
              id: 'stagger-attributes-incident_stagger-2_alpha',
              position: 0,
              duration: 800,
            },
          },
        },
        'stagger-attributes-incident_stagger-3': {
          id: 'stagger-attributes-incident_stagger-3',
          position: 0,
          duration: 800,
          children: {
            'stagger-attributes-incident_stagger-3_alpha': {
              id: 'stagger-attributes-incident_stagger-3_alpha',
              position: 0,
              duration: 800,
            },
          },
        },
        'stagger-attributes-incident_stagger-4': {
          id: 'stagger-attributes-incident_stagger-4',
          position: 0,
          duration: 800,
          children: {
            'stagger-attributes-incident_stagger-4_alpha': {
              id: 'stagger-attributes-incident_stagger-4_alpha',
              position: 0,
              duration: 800,
            },
          },
        },
        'stagger-attributes-incident_stagger-5': {
          id: 'stagger-attributes-incident_stagger-5',
          position: 0,
          duration: 800,
          children: {
            'stagger-attributes-incident_stagger-5_alpha': {
              id: 'stagger-attributes-incident_stagger-5_alpha',
              position: 0,
              duration: 800,
            },
          },
        },
        'stagger-attributes-incident_stagger-6': {
          id: 'stagger-attributes-incident_stagger-6',
          position: 0,
          duration: 800,
          children: {
            'stagger-attributes-incident_stagger-6_alpha': {
              id: 'stagger-attributes-incident_stagger-6_alpha',
              position: 0,
              duration: 800,
            },
          },
        },
      },
    },
    'test-group-1': {
      id: 'test-group-1',
      position: 1000,
      duration: 9000,
      children: {
        'test-incident-1': {
          id: 'test-incident-1',
          position: 7000,
          duration: 2000,
          children: {
            'test-incident-1_my-class-1': {
              id: 'test-incident-1_my-class-1',
              position: 0,
              duration: 2000,
              children: {
                'test-incident-1_my-class-1_alpha': {
                  id: 'test-incident-1_my-class-1_alpha',
                  position: 0,
                  duration: 2000,
                },
                'test-incident-1_my-class-1_beta': {
                  id: 'test-incident-1_my-class-1_beta',
                  position: 0,
                  duration: 2000,
                },
              },
            },
            'test-incident-1_my-class-2': {
              id: 'test-incident-1_my-class-2',
              position: 0,
              duration: 2000,
              children: {
                'test-incident-1_my-class-2_alpha': {
                  id: 'test-incident-1_my-class-2_alpha',
                  position: 0,
                  duration: 2000,
                },
                'test-incident-1_my-class-2_beta': {
                  id: 'test-incident-1_my-class-2_beta',
                  position: 0,
                  duration: 2000,
                },
              },
            },
          },
        },
      },
    },
  },
};
