// @donkeyclip/motorcortex/src/main.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import BrowserClip from './Clip/Browser/BrowserClip';
import DescriptiveAudioClip from './Clip/Browser/Descriptive/DescriptiveAudioClip';
import HTMLDescriptiveClip from './Clip/Browser/Descriptive/HTMLDescriptiveClip';
import audioPlugin from './Clip/WebAudio';
import { CSSEffect, setCore as setCSSCore } from './CSSLayer/CSSLayer';
import ExtendableCSSEffect from './CSSLayer/ExtendableCSSEffect/ExtendableCSSEffect';
import DescriptiveCombo from './DescriptiveIncidents/DescriptiveCombo';
import DescriptiveGroup from './DescriptiveIncidents/DescriptiveGroup';
import clipFromDefinition from './IncidentFromDefinition/clipFromDefinition';
import MediaPlayback from './RealIncidents/MediaPlayIncident';
import MonoIncident from './RealIncidents/MonoIncident';
import easings from './_configuration/easings';
import { loadPlugin } from './_coreUtils/pluginLoader';
import TimeCapsule from './_coreUtils/TimeCapsule';
import version from './version';

export const utils = {
  easings,
  clipFromDefinition,
};

export const {
  AudioEffect,
  AudioPlayback,
  Clip: CoreAudioClip,
} = loadPlugin(audioPlugin);

export {
  BrowserClip,
  CSSEffect,
  DescriptiveAudioClip as AudioClip,
  HTMLDescriptiveClip as HTMLClip,
  DescriptiveCombo as Combo,
  DescriptiveGroup as Group,
  MediaPlayback,
  MonoIncident as Effect,
  TimeCapsule,
  setCSSCore,
  loadPlugin,
  version,
  ExtendableCSSEffect,
};
